cmd/clusterctl/client/tree/annotations.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tree
+
+import (
+	"strconv"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ShowObjectConditionsAnnotation documents that the presentation layer should show all the conditions for the object.
+	ShowObjectConditionsAnnotation = "tree.cluster.x-k8s.io.io/show-conditions"
+
+	// ObjectMetaNameAnnotation contains the meta name that should be used for the object in the presentation layer,
+	// e.g. control plane for KCP.
+	ObjectMetaNameAnnotation = "tree.cluster.x-k8s.io.io/meta-name"
+
+	// VirtualObjectAnnotation documents that the object does not correspond to any real object, but instead is
+	// a virtual object introduced to provide a better representation of the cluster status, e.g. workers.
+	VirtualObjectAnnotation = "tree.cluster.x-k8s.io.io/virtual-object"
+
+	// GroupingObjectAnnotation documents that the child of this node will be grouped in case the ready condition
+	// has the same Status, Severity and Reason.
+	GroupingObjectAnnotation = "tree.cluster.x-k8s.io.io/grouping-object"
+
+	// GroupObjectAnnotation documents that the object does not correspond to any real object, but instead it is
+	// a virtual node representing group of sibling nodes, e.g. a group of machines.
+	GroupObjectAnnotation = "tree.cluster.x-k8s.io.io/group-object"
```

Comment:
As I was going through the code, I noticed that it was hard to keep track of `Grouping` vs `Group` since the words are quite similar. 
Could we perhaps rename the `GroupingObjectAnnotation` to something like  
`SameReadyConditionAnnotation = "tree.cluster.x-k8s.io.io/ready-condition-group"`
or 
`GroupByReadyConditionAnnotation = "tree.cluster.x-k8s.io.io/ready-condition-group"`?

Reply:
Uhhm, I'm not sure this name makes it clear the node is responsible for grouping the nodes.

Given that the grouping annotation is not intended for the end-users, but it triggers an internal behavior of this library I'm leaning towards keeping things as they are, and eventually send a separated PR to make the annotation internal/remove it from the nodes after processing is completed.

Reply:
Hmm..I suggested the rename since the grouping is happening specifically based on the ready condition. This is another option. `GroupedByReadyConditionAnnotation = "tree.cluster.x-k8s.io.io/grouped-by-ready-object"`

Perhaps, the other one could be  `VirtualGroupObjectAnnotation = "tree.cluster.x-k8s.io.io/virtual-group-object"`.

I think it makes sense for these annotations to be internal since it won't be used by the clusterctl library users. However, this suggestion was more from a CAPI developer perspective rather than a library user. As a CAPI developer it was a little difficult for me to immediately understand what `GroupingObjectAnnotation` vs `GroupObjectAnnotation`

Reply:
I'm happy to talk about this comment, but TBH I can't grasp what are you aiming at (so my comment below might be somehow out of focus).

GroupingObjectAnnotation is an annotation that tells to the library that a node should group its own children, but I don't see value addend in surfacing how grouping is done; also, if I look at this the other way around, this could be brittle in case we want to change the grouping rule in the future.

The proposed VirtualGroupObjectAnnotation = "tree.cluster.x-k8s.io.io/virtual-group-object" reminds me of a mix of node attribute (the node is virtual) and the origin of the node (this node was originated by a grouping operation); Instead I think it is better to keep the two separated into two annotations, as we are already doing today so things are more consistent with other virtual nodes

Reply:
@wfernandes pointed out that the `GroupingObjectAnnotation` specifies in the comment that it uses the `Ready` condition to group objects, which the text doesn't really surface. That makes sense to me, I'm also inclined to use `GroupBy` if possible, which is a familiar concept coming from SQL databases

Reply:
@vincepri @wfernandes happy to talk about it! 
I'm already using "group" in the annotation names, with the two variant "grouping" and "group" according to where the annotation applies, so it should not be a big change.
But as I said, I'm not sure to understand where this should lead to.

Reply:
It just makes the name more explicit, which seems ok to me?

Reply:
@vincepri I got we are discussing a rename, but I really would like to understand the nuances between the current "grouping" (for tagging the node responsibile to group its children), "group" (for tagging the node resulting from the group action), and the proposed variants, which, as I explained above, IMO are mixing different concerns that in some cases are already tracked by other annotations (the type of the resulting node, virtual, or the implementation details on how grouping happens).

We should also consider that these names currently are consistent with the library methods, and this is something I would like to preserve no matter the final name we should agree on; e.g.g as of today library user's are using:
```
tree.Add(..., GroupingObject(true))
```
or
```
if tree.IsGroupObject(obj) {
   ...
}
```
without being forced to know by hearth/hard code annotations names

Also `doc.go` is using the same wording...

Reply:
As agreed with @wfernandes I'm going to improve annotation comments, but given that everything is hidden by the library API, we can leave the current annotation names

Reply:
@wfernandes PTAL; changes to comments are in a separated commit for easier review